rawfile/utils/snapshot_manager.py

```diff
@@ -0,0 +1,115 @@
+from pathlib import Path
+import time
+from typing import List
+from dataclasses import dataclass
+import consts
+from utils.commands import run
+
+from utils.devices import device_to_mountpoint
+from utils.lock import VolLock
+from utils.rawfile import attached_loops, img_file, snapshots_dir
+from glob import glob
+
+
+@dataclass
+class Snapshot:
+    name: str
+    volume_id: str
+    size_bytes: int
+    creation_time: float
+    ready: bool = True
+
+
+class SnapshotManager:
+    def _get_snapshot_path(self, volume_id: str, name: str) -> Path:
+        return snapshots_dir(volume_id) / f"{name}.img"
+
+    def create_snapshot(
+        self,
+        volume_id: str,
+        name: str,
+        copy_on_write: bool = False,
+        freeze_fs: bool = False,
+    ) -> Snapshot:
+        with VolLock(volume_id):
+            file = img_file(volume_id)
+            loop_devs = attached_loops(file.as_posix())
+            snap_path = self._get_snapshot_path(volume_id, name)
+            snap_path.parent.mkdir(parents=True, exist_ok=True)
+            Path(f"{snap_path}.creating").touch()
+            Path(snap_path).unlink(missing_ok=True)
+            mountpoints = []
+            snap_inuse = freeze_fs or copy_on_write
+            if (loop_devs and len(loop_devs)) and (not snap_inuse):
+                raise  # FIXME: Raise a better Exception
+            if loop_devs and len(loop_devs):
+                for dev in loop_devs:
+                    mp = device_to_mountpoint(dev, first_only=False)
+                    if mp:
+                        mountpoints.extend(mp)
+            try:
+                if freeze_fs and mountpoints:
+                    for mp in mountpoints:
+                        run(f"fsfreeze --freeze {mp}", check=True)
+
+                reflink = "always" if copy_on_write else "never"
+                cmd = f"cp --sparse=always --reflink={reflink} {file} {snap_path}"
```

Comment:
btw seems cp returns not supported with reflink=always if the fs doesn't support it, could we make use of this to determine cow?

Reply:
If that's the case we can do a fake cp command to determine that on startup, and save it into memory

Reply:
yes that would be neat :)

Reply:
I will make it to determine that by default, but users are able to override that

Reply:
Done.